### PR TITLE
all done

### PR DESCRIPTION
**Pull request by vlykhvar**
> CinemaHallController

1. Add cinema hall - POST: /cinema-halls
2. Get all cinema halls - GET: /cinema-halls

> MovieController

1. Add a movie - POST: /movies
2. Get all movies - GET: /movies

> MovieSessionController

1. Add movie session - POST: /movie-sessions
2. Get all available movie sessions - GET: /movie-sessions/available?movieId=1&date=29.02.2020
3. Update movie session - PUT /movie-sessions
4. Remove movie session - DELETE /movie-sessions

### CODE REVIEW

**Review by Ketch92 — APPROVED** [review]

File: src/main/java/com/dev/theater/controller/CinemaHallController.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    @GetMapping
+    public List<CinemaHallResponseDto> getAllCinemaHall() {
+        return cinemaHallService.getAll().stream()
+                .map(cinemaHallMapper::cinemaHallToDto)
+                .collect(Collectors.toList());
+    }
+
+    @PostMapping
+    public void addCinemaHall(CinemaHallResponseDto cinemaHallDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/controller/CinemaHallController.java]: why responseDto here?

File: src/main/java/com/dev/theater/controller/MovieController.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.stream.Collectors;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/movie")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/controller/MovieController.java]:
```suggestion
@RequestMapping("/movies")
```

File: src/main/java/com/dev/theater/controller/MovieController.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    @GetMapping
+    public List<MovieResponseDto> getMovies() {
+        return movieService.getAll().stream()
+                .map(movieMapper::movieToDto)
+                .collect(Collectors.toList());
+    }
+
+    @PostMapping
+    public void addMovie(@RequestBody MovieRequestDto movieRequestDtoDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/controller/MovieController.java]:
```suggestion
    public void addMovie(@RequestBody MovieRequestDto movieRequestDto) {
```

File: src/main/java/com/dev/theater/controller/MovieSessionController.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/moviesessions")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/controller/MovieSessionController.java]: see task description - what the mapping should be?

File: src/main/java/com/dev/theater/controller/MovieSessionController.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Autowired
+    public MovieSessionController(MovieSessionMapper movieSessionMapper,
+                                  MovieSessionService movieSessionService) {
+        this.movieSessionMapper = movieSessionMapper;
+        this.movieSessionService = movieSessionService;
+    }
+
+    @PostMapping
+    public void addMovieSession(@RequestBody MovieSessionRequestDto movieSessionResponseDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/controller/MovieSessionController.java]:
```suggestion
    public void addMovieSession(@RequestBody MovieSessionRequestDto movieSessionRequestDto) {
```

File: src/main/java/com/dev/theater/mapper/CinemaHallMapper.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+public class CinemaHallMapper {
+    public CinemaHallResponseDto cinemaHallToDto(CinemaHall cinemaHall) {
+        CinemaHallResponseDto cinemaHallDto = new CinemaHallResponseDto();
+        cinemaHallDto.setCapacity(cinemaHall.getCapacity());
+        cinemaHallDto.setId(cinemaHall.getId());
+        cinemaHallDto.setDescription(cinemaHall.getDescription());
+        return cinemaHallDto;
+    }
+
+    public CinemaHall dtoToCinemaHall(CinemaHallResponseDto cinemaHallDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/mapper/CinemaHallMapper.java]: why responseDto?

File: src/main/java/com/dev/theater/mapper/MovieSessionMapper.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        movieSessionDto.setMovieTitle(movieSession.getMovie().getTitle());
+        movieSessionDto.setCinemaHallId(movieSession.getCinemaHall().getId());
+        movieSessionDto.setCinemaHallDescription(movieSession.getCinemaHall().getDescription());
+        return movieSessionDto;
+    }
+
+    public MovieSession dtoToMovieSession(MovieSessionRequestDto movieSessionRequestDto) {
+        MovieSession movieSession = new MovieSession();
+        movieSession.setMovie(new Movie());
+        movieSession.setCinemaHall(new CinemaHall());
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/mapper/MovieSessionMapper.java]: we should get movieId and cinemaHallId in the movieSessionRequestDto

then we should fetch entities by their ids through `movieService` and `cinemaHallService`
and set them to the movieSession

File: src/main/java/com/dev/theater/model/dto/MovieSessionRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
+package com.dev.theater.model.dto;
+
+public class MovieSessionRequestDto {
+    private String movieTitle;
+    private String showTime;
+    private Long cinemaHallId;
+    private String cinemaHallDescription;
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/model/dto/MovieSessionRequestDto.java]: the only informative fields that we can get from this dto are:
`movieId`
`cinemaHallId`
`showTime`

let's refactor the fields

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/theater/controller/CinemaHallController.java [hunk 9]
```diff
@@ -33,8 +34,8 @@ public CinemaHallController(CinemaHallMapper cinemaHallMapper,
     }
 
     @PostMapping
-    public void addCinemaHall(CinemaHallResponseDto cinemaHallDto) {
-        CinemaHall cinemaHall = cinemaHallMapper.dtoToCinemaHall(cinemaHallDto);
+    public void addCinemaHall(CinemaHallRequestDto cinemaHallRequestDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/theater/controller/CinemaHallController.java]: let's test this method and see what may be missing  comparing to POST method in other controllers

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**